contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -427,11 +427,36 @@ ensure that GETs of individual objects remain bounded in time and space, these
 sets may be queried via separate API queries, but will not be expanded in the
 referring object's status.
 
-References to specific objects, especially specific resource versions and/or
-specific fields of those objects, are specified using the `ObjectReference` type
-(or other types representing strict subsets of it). Unlike partial URLs, the
-ObjectReference type facilitates flexible defaulting of fields from the
-referring object or other contextual information.
+References to specific objects are specified using object reference types that
+uniquely identify their referents.  Object reference types may be defined along
+with the referring object types.  An object reference type must have a required
+`name` field.  In addition, the type may have a `namespace` field to accommodate
+references to non-local objects (that is, namespaced referents that are not in
+the same namespace as the referring object).  If the reference needs to
+accommodate referents of different resource types, the object reference type may
+also have `group` and `resource` fields.  For example:
+
+```go
+// FooReference holds a reference to an API object that can be used as a foo.
+type FooReference struct {
+	// group is the group of the referent.  The empty string represents
+	// the core API group.
+	Group string
+	// resource is the resource of the referent.
+	Resource string
+	// namespace is the namespace of the referent.
+	Namespace string
+	// name is the name of the referent.
+	Name string
+}
+```
+
+Unlike partial URLs, object reference types facilitate flexible defaulting of
+fields from the referring object or other contextual information.
+
+Some existing APIs use the generic `ObjectReference`, `LocalObjectReference`,
```

Comment:
@liggitt We're having difficulty understanding the current recommended best practices here. The [guidance added to ObjectReference](https://github.com/kubernetes/kubernetes/pull/87459) by @deads2k suggests the following:

> Instead of using this type, create a locally provided and used type that is well-focused on your reference.

At the same time, it sounds like you're suggesting that `TypedLocalObjectReference` might still be worth using here? 

This is in the context of a new set of Service APIs that uses _lots_ of object references. So far we've relied on the guidance from that comment on ObjectReference and created our object reference types for all the things. Would it be preferable to reuse the core type?

We're also having trouble understanding the suggestion added in that same PR to use Resource instead of Kind in object references. That seems like it would be especially confusing for users that are used to Kind in object refs.

Right now we're in a confusing state where the godocs seem to disagree with the API conventions and it's hard to know which to follow.

Reply:
> So far we've relied on the guidance from that comment on ObjectReference and created custom object reference types for all the things. Would it be preferable to reuse the core type?

A scoped type with exactly the fields you need is a good idea. LocalObjectReference is so simplistic as to make reuse almost more trouble than it's worth. TypedLocalObjectReference is more constrained than ObjectReference, but reusing the shared struct means you can't put any context-specific documentation/descriptions on the generated API doc.

> We're also having trouble understanding the suggestion added in that same PR to use Resource instead of Kind in object references. That seems like it would be especially confusing for users that are used to Kind in object refs.

Typically, references are used to look up other objects via API calls. To make those API calls, you always need to resolve to a resource (which shows up in the API request URL path). You _can_ use discovery to resolve a Kind to a resource, but since multiple resources _could_ accept the same Kind, you would have to have a convention of taking the first resource associated with the Kind. Using the resource directly in the reference is less ambiguous.

Reply:
> You can use discovery to resolve a Kind to a resource, but since multiple resources could accept the same Kind, you would have to have a convention of taking the first resource associated with the Kind. Using the resource directly in the reference is less ambiguous.

Isn't this solved by the introduction of the `Group`? The convention I am used to as a controller developer is that a `Resource` translates to `<kind>/<name>`, but that does not seem to apply here as there also is a `Name` field in the example structure. Given this, I think convention like the following would be more intuitive to users _and_ developers:

 ```go
// FooReference holds a reference to an API object that can be used as a foo.
type FooReference struct {
	// Group is the group of the referent. The empty string represents
	// the core API group.
	Group string
	// Kind is the kind of of the referent, and must exist in the referenced Group.
	Kind string
	// Namespace is the namespace of the referent.
	Namespace string
	// Name is the name of the referent.
	Name string
}

This would also match what is currently defined in `CustomResourceDefinition` resources:

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: secrets.secret.example.com
spec:
  group: secret.example.com
  names:
    kind: Secret
    listKind: SecretList
    plural: secrets
    singular: secret
```

which would translate to:

```go
&FooReference{
	Group:     "secret.example.com",
	Kind:      "Secret",
	Namespace: "default",
	Name:      "foo",
}
```

while not causing conflicts with the core `Secret` resource kind.